grafanaplane/oncall/schedule.libsonnet

```diff
@@ -0,0 +1,45 @@
+local d = import 'github.com/jsonnet-libs/docsonnet/doc-util/main.libsonnet';
+local xtd = import 'github.com/jsonnet-libs/xtd/main.libsonnet';
+
+local util = import '../util.libsonnet';
+local raw = import '../zz/main.libsonnet';
+local schedule = raw.oncall.v1alpha1.schedule;
+local forProvider = schedule.spec.parameters.forProvider;
+
+forProvider  // raise forProvider functions to here
+{
+  local base = self,
+
+  '#new':: d.func.new(
+    |||
+      `new` creates a Schedule. The `name` is a display-friendly
+      string, and `id` defaults to a slug-ified version of it.
+    |||,
+    [
+      d.argument.new('name', d.T.string),
+      d.argument.new('type', d.T.string, enums=['ical', 'calendar']),
+      d.argument.new('providerName', d.T.string),
+      d.argument.new('resourceId', d.T.string, default='rfc1123(name)'),
+    ]
+  ),
+  new(name, id=xtd.ascii.stringToRFC1123(name))::
+    schedule.new(id)
+    + schedule.spec.parameters.forProvider.withName(name),
+
+  '#withShifts':: d.func.new(
+    |||
+      `withShifts` sets a Schedule to type `calendar` and configures shifts.
+      Shifts are only applicable to `calendar` type Schedules. `shifts` is an
+      array of Shift resource names or entire Shift manifests.
+    |||,
+    [
+      d.argument.new('shifts', d.T.array, default='null'),
+    ]
+  ),
+  withShifts(shifts=null)::
+    super.withType('calendar')
+    + super.withShiftsRef([
+      super.shiftsRef.withName(util.getName(shift))
+      for shift in shifts
+    ]),
```

Comment:
I'd make this an integrated library as well, and perhaps nest it in a 'calendar' key so we can create entry points for each type of schedule.

```jsonnet
{
  calendar: {
    new(name, id=xtd.ascii.stringToRFC1123(name)):: {
      shedule:
        schedule.new(id)
        + schedule.spec.parameters.forProvider.withName(name)
        + schedule.spec.parameters.forProvider.withType('calendar')
        + schedule.spec.parameters.forProvider.withShiftsRef([
          schedule.spec.parameters.forProvider.shiftsRef.withName(shift.metadata.name)
          for shift in self.shifts
        ]),

      shifts: [],
    },

    withShifts(shifts): {
      shifts: shifts,
    },
  },
}
```

Reply:
I just applied this, and opted to move Shifts fully underneath this `calendar` object, since this is the only context they're used in. I opted to take them as an object, rather than an array, since they're not ordered in any meaningful way. I also included an example of how I'm consuming the API this way, which I'm hoping is fairly illustrative.